File: Iverson.Clients/TypeScript/src/annotations.ts
/**
 * Decorator-based annotation system for Iverson entities.
 *
 * Usage:
 *   @IversonEntity()
 *   class Article {
 *     @IversonKey()
 *     id: string = '';
 *
 *     @IversonLargeField()
 *     body: string = '';
 *
 *     @IversonSearchKey(0)
 *     category: string = '';
 *
 *     @ManyToOne(() => Author)
 *     authorId: string = '';
 *   }
 */
import 'reflect-metadata';

import { ClrType } from '../generated/object_mapping.js';

// ── Metadata symbol keys ───────────────────────────────────────────────────────

const IVERSON_ENTITY_KEY   = Symbol('iverson:entity');
const IVERSON_KEY_KEY      = Symbol('iverson:key');
const IVERSON_SEARCH_KEYS  = Symbol('iverson:search_keys');
const IVERSON_LARGE_FIELDS = Symbol('iverson:large_fields');
const IVERSON_EMBEDDING_FIELDS = Symbol('iverson:embedding_fields');
const IVERSON_CHUNK_FIELDS     = Symbol('iverson:chunk_fields');
const IVERSON_RELATIONS    = Symbol('iverson:relations');
const IVERSON_METADATA_FIELDS  = Symbol('iverson:metadata_fields');
const IVERSON_SUMMARY_FIELDS   = Symbol('iverson:summary_fields');
const IVERSON_KEYWORDS_FIELDS  = Symbol('iverson:keywords_fields');
const IVERSON_EXTRACTED_FIELDS = Symbol('iverson:extracted_fields');
const IVERSON_PROPERTY_DESCRIPTIONS = Symbol('iverson:property_descriptions');
const IVERSON_TYPE_DESCRIPTION      = Symbol('iverson:type_description');
const IVERSON_TENANT_FIELDS    = Symbol('iverson:tenant_fields');

// ── Public relation kind constants ─────────────────────────────────────────────

export type RelationKindString = 'many_to_one' | 'many_to_many' | 'one_to_many' | 'one_to_one';

export interface RelationMeta {
    field: string;
    kind: RelationKindString;
    relatedType: string;
}

export interface SearchKeyMeta {
    field: string;
    order: number;
}

// ── @IversonEntity() ───────────────────────────────────────────────────────────

export function IversonEntity(): ClassDecorator {
    return (target) => {
        Reflect.defineMetadata(IVERSON_ENTITY_KEY, true, target);
    };
}

export function isIversonEntity(target: Function): boolean {
    return Reflect.getMetadata(IVERSON_ENTITY_KEY, target) === true;
}

// ── @IversonKey() ──────────────────────────────────────────────────────────────

export function IversonKey(): PropertyDecorator {
    return (target, propertyKey) => {
        Reflect.defineMetadata(IVERSON_KEY_KEY, String(propertyKey), target.constructor);
    };
}

export function getKeyField(target: Function): string | undefined {
    return Reflect.getMetadata(IVERSON_KEY_KEY, target);
}

// ── @IversonSearchKey(order) ───────────────────────────────────────────────────

export function IversonSearchKey(order: number): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: SearchKeyMeta[] =
            Reflect.getMetadata(IVERSON_SEARCH_KEYS, target.constructor) ?? [];
        existing.push({ field: String(propertyKey), order });
        Reflect.defineMetadata(IVERSON_SEARCH_KEYS, existing, target.constructor);
    };
}

export function getSearchKeys(target: Function): SearchKeyMeta[] {
    const keys: SearchKeyMeta[] = Reflect.getMetadata(IVERSON_SEARCH_KEYS, target) ?? [];
    return [...keys].sort((a, b) => a.order - b.order);
}

// ── @IversonLargeField() ──────────────────────────────────────────────────────

export function IversonLargeField(): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: string[] =
            Reflect.getMetadata(IVERSON_LARGE_FIELDS, target.constructor) ?? [];
        existing.push(String(propertyKey));
        Reflect.defineMetadata(IVERSON_LARGE_FIELDS, existing, target.constructor);
    };
}

export function getLargeFields(target: Function): string[] {
    return Reflect.getMetadata(IVERSON_LARGE_FIELDS, target) ?? [];
}

// ── @IversonEmbedding() ──────────────────────────────────────────────────────

export function IversonEmbedding(): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: string[] =
            Reflect.getMetadata(IVERSON_EMBEDDING_FIELDS, target.constructor) ?? [];
        existing.push(String(propertyKey));
        Reflect.defineMetadata(IVERSON_EMBEDDING_FIELDS, existing, target.constructor);
    };
}

export function getEmbeddingFields(target: Function): string[] {
    return Reflect.getMetadata(IVERSON_EMBEDDING_FIELDS, target) ?? [];
}

// ── @IversonChunk(maxTokens, overlap) ────────────────────────────────────────

export interface ChunkMeta {
    field: string;
    maxTokens: number;
    overlap: number;
    contextual: boolean;
}

export interface ChunkOptions {
    contextual?: boolean;
}

export function IversonChunk(maxTokens: number = 512, overlap: number = 64, options: ChunkOptions = {}): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: ChunkMeta[] =
            Reflect.getMetadata(IVERSON_CHUNK_FIELDS, target.constructor) ?? [];
        existing.push({ field: String(propertyKey), maxTokens, overlap, contextual: options.contextual ?? false });
        Reflect.defineMetadata(IVERSON_CHUNK_FIELDS, existing, target.constructor);
    };
}

export function getChunkFields(target: Function): ChunkMeta[] {
    return Reflect.getMetadata(IVERSON_CHUNK_FIELDS, target) ?? [];
}

// ── @IversonSummary() ─────────────────────────────────────────────────────────

/** Marks a property as the target for an Ollama-driven summary during ingest enrichment. */
export function IversonSummary(): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: string[] =
            Reflect.getMetadata(IVERSON_SUMMARY_FIELDS, target.constructor) ?? [];
        existing.push(String(propertyKey));
        Reflect.defineMetadata(IVERSON_SUMMARY_FIELDS, existing, target.constructor);
    };
}

export function getSummaryFields(target: Function): string[] {
    return Reflect.getMetadata(IVERSON_SUMMARY_FIELDS, target) ?? [];
}

// ── @IversonKeywords() ────────────────────────────────────────────────────────

/** Marks a property as the target for Ollama-driven keyword extraction during ingest enrichment. */
export function IversonKeywords(): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: string[] =
            Reflect.getMetadata(IVERSON_KEYWORDS_FIELDS, target.constructor) ?? [];
        existing.push(String(propertyKey));
        Reflect.defineMetadata(IVERSON_KEYWORDS_FIELDS, existing, target.constructor);
    };
}

export function getKeywordsFields(target: Function): string[] {
    return Reflect.getMetadata(IVERSON_KEYWORDS_FIELDS, target) ?? [];
}

// ── @IversonExtracted(hint) ───────────────────────────────────────────────────

export interface ExtractedMeta {
    field: string;
    hint: string;
}

/**
 * Marks a property as the target for an Ollama-driven extraction during
 * ingest enrichment, guided by `hint`.
 *
 * The hint is mandatory: the server only treats a property as an extraction
 * target when a non-empty hint is present (`SchemaBuilder.cs` only creates
 * the Extracted target when the hint is non-empty), so a blank hint would be
 * silently dropped server-side. This decorator rejects that case up front —
 * at decoration time, not just at the type level, since JS callers bypass
 * TypeScript's compile-time checks entirely.
 */
export function IversonExtracted(hint: string): PropertyDecorator {
    return (target, propertyKey) => {
        if (hint === undefined || hint === null || hint.trim() === '') {
            throw new Error(
                `@IversonExtracted() on ${target.constructor.name}.${String(propertyKey)} requires a ` +
                'non-blank extraction hint; the server treats an empty hint as "not an extraction ' +
                'target" and would silently drop it.',
            );
        }
        const existing: ExtractedMeta[] =
            Reflect.getMetadata(IVERSON_EXTRACTED_FIELDS, target.constructor) ?? [];
        existing.push({ field: String(propertyKey), hint });
        Reflect.defineMetadata(IVERSON_EXTRACTED_FIELDS, existing, target.constructor);
    };
}

export function getExtractedFields(target: Function): ExtractedMeta[] {
    return Reflect.getMetadata(IVERSON_EXTRACTED_FIELDS, target) ?? [];
}

// ── @IversonMetadata() ───────────────────────────────────────────────────────

/** Marks a scalar property as chunk metadata — denormalized onto chunk points. */
export function IversonMetadata(): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: string[] =
            Reflect.getMetadata(IVERSON_METADATA_FIELDS, target.constructor) ?? [];
        existing.push(String(propertyKey));
        Reflect.defineMetadata(IVERSON_METADATA_FIELDS, existing, target.constructor);
    };
}

export function getMetadataFields(target: Function): string[] {
    return Reflect.getMetadata(IVERSON_METADATA_FIELDS, target) ?? [];
}

// ── @IversonArray(elementType) ─────────────────────────────────────────────────

const IVERSON_ARRAY_KEY = Symbol('iverson:array');

/**
 * Declares a property as an array column, naming its element type explicitly.
 * TypeScript cannot infer the element type: `emitDecoratorMetadata` erases it
 * (design:type reports only the `Array` constructor), and an initialized `[]`
 * carries no element type either. Without this decorator, the registrar would
 * either silently fall back to a scalar CLR_STRING column or, for undecorated
 * declarations, skip detection entirely.
 */
export function IversonArray(elementType: ClrType): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: Map<string, ClrType> =
            Reflect.getMetadata(IVERSON_ARRAY_KEY, target.constructor) ?? new Map();
        existing.set(String(propertyKey), elementType);
        Reflect.defineMetadata(IVERSON_ARRAY_KEY, existing, target.constructor);
    };
}

export function getArrayFields(target: Function): Map<string, ClrType> {
    return Reflect.getMetadata(IVERSON_ARRAY_KEY, target) ?? new Map();
}

// ── @IversonGuid() ──────────────────────────────────────────────────────────

const IVERSON_GUID_KEY = Symbol('iverson:guid');

/**
 * Declares a property as a UUID column. TypeScript has no UUID type — a GUID is
 * carried as a `string` — so the runtime cannot distinguish it from any other
 * string. The server requires key and foreign-key columns to be UUID.
 */
export function IversonGuid(): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: Set<string> =
            Reflect.getMetadata(IVERSON_GUID_KEY, target.constructor) ?? new Set();
        existing.add(String(propertyKey));
        Reflect.defineMetadata(IVERSON_GUID_KEY, existing, target.constructor);
    };
}

export function getGuidFields(target: Function): Set<string> {
    return Reflect.getMetadata(IVERSON_GUID_KEY, target) ?? new Set();
}

// ── @IversonTenant() ────────────────────────────────────────────────────────

/**
 * Marks the scalar property that carries the tenant identifier. Exactly one
 * property per entity must carry this decorator: the server rejects schema
 * registration when `tenant_field` is missing, and this decorator's own
 * marked-property count is validated in `describeEntity` (zero or more than
 * one is a client-side error, since the server only ever sees a single name
 * on the wire and cannot detect that duplication itself).
 */
export function IversonTenant(): PropertyDecorator {
    return (target, propertyKey) => {
        const existing: string[] =
            Reflect.getMetadata(IVERSON_TENANT_FIELDS, target.constructor) ?? [];
        existing.push(String(propertyKey));
        Reflect.defineMetadata(IVERSON_TENANT_FIELDS, existing, target.constructor);
    };
}

export function getTenantFields(target: Function): string[] {
    return Reflect.getMetadata(IVERSON_TENANT_FIELDS, target) ?? [];
}

// ── @IversonDescription(text) ────────────────────────────────────────────────

/**
 * Attaches free-form descriptive text. Usable on an entity class (populates
 * TypeDescriptor.description) or on any property (populates
 * PropertyDescriptor.description), including the key property.
 */
export function IversonDescription(text: string): ClassDecorator & PropertyDecorator {
    return ((target: any, propertyKey?: string | symbol) => {
        if (propertyKey === undefined) {
            // Class decorator: target is the constructor.
            Reflect.defineMetadata(IVERSON_TYPE_DESCRIPTION, text, target);
            return;
        }
        const ctor = target.constructor;
        const existing: Record<string, string> =
            Reflect.getMetadata(IVERSON_PROPERTY_DESCRIPTIONS, ctor) ?? {};
        existing[String(propertyKey)] = text;
        Reflect.defineMetadata(IVERSON_PROPERTY_DESCRIPTIONS, existing, ctor);
    }) as ClassDecorator & PropertyDecorator;
}

export function getTypeDescription(target: Function): string {
    return Reflect.getMetadata(IVERSON_TYPE_DESCRIPTION, target) ?? '';
}

export function getPropertyDescriptions(target: Function): Record<string, string> {
    return Reflect.getMetadata(IVERSON_PROPERTY_DESCRIPTIONS, target) ?? {};
}

// ── Relation decorators ────────────────────────────────────────────────────────

export interface PendingRelationMeta {
    field: string;
    kind: RelationKindString;
    typeFactory: () => Function;
}

function addRelation(target: object, propertyKey: string | symbol, kind: RelationKindString, typeFactory: () => Function): void {
    const ctor = (target as any).constructor;
    const existing: PendingRelationMeta[] =
        Reflect.getMetadata(IVERSON_RELATIONS, ctor) ?? [];
    existing.push({ field: String(propertyKey), kind, typeFactory });
    Reflect.defineMetadata(IVERSON_RELATIONS, existing, ctor);
}

/**
 * Each of these decorators stores the raw `typeFactory` rather than calling it
 * immediately. Property decorators run at class-definition time, and calling
 * the factory eagerly would dereference the related class before it exists
 * for any forward reference (including the two halves of a genuine mutual
 * reference, where reordering the classes cannot help either direction).
 * Resolution is deferred to `getRelations`, which runs at schema-registration
 * time after every class in the module has finished initializing.
 */
export function ManyToOne(typeFactory: () => Function): PropertyDecorator {
    return (target, propertyKey) => {
        addRelation(target, propertyKey, 'many_to_one', typeFactory);
    };
}

export function ManyToMany(typeFactory: () => Function): PropertyDecorator {
    return (target, propertyKey) => {
        addRelation(target, propertyKey, 'many_to_many', typeFactory);
    };
}

export function OneToMany(typeFactory: () => Function): PropertyDecorator {
    return (target, propertyKey) => {
        addRelation(target, propertyKey, 'one_to_many', typeFactory);
    };
}

export function OneToOne(typeFactory: () => Function): PropertyDecorator {
    return (target, propertyKey) => {
        addRelation(target, propertyKey, 'one_to_one', typeFactory);
    };
}

export function getRelations(target: Function): RelationMeta[] {
    const pending: PendingRelationMeta[] = Reflect.getMetadata(IVERSON_RELATIONS, target) ?? [];
    return pending.map(({ field, kind, typeFactory }) => ({
        field,
        kind,
        relatedType: typeFactory().name,
    }));
}

/**
 * Like `getRelations`, but returns the raw, unresolved `typeFactory` instead of collapsing it
 * to a name. `getRelations` is public API with production and test call sites that depend on
 * its name-only shape, so it is left untouched; the read path needs the actual related class
 * (not just its name) to construct and recursively hydrate a typed instance, and the only place
 * that can hand that out is here, since `IVERSON_RELATIONS` is module-private to this file.
 */
export function getRelationsWithFactory(target: Function): PendingRelationMeta[] {
    return Reflect.getMetadata(IVERSON_RELATIONS, target) ?? [];
}

File: Iverson.Clients/Python/iverson_client/annotations.py
"""
Decorator-based annotation system for Iverson entities.

Usage:
    @iverson_entity
    class Article:
        id: str = iverson_key()
        title: str = None
        body: str = iverson_large_field()
        category: str = iverson_search_key(order=0)
        published_at: datetime = iverson_search_key(order=1)
        author_id: str = many_to_one('Author')
        transcript: str = iverson_field(large_field=True, chunk=True)
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(kw_only=True)
class FieldMeta:
    """Descriptor object attached as a class attribute default.

    ``@iverson_entity`` inspects ``__annotations__`` and replaces every
    ``FieldMeta`` default with ``None`` so the attribute behaves normally
    at runtime while preserving the metadata on ``cls._iverson_meta``.

    Every scalar declaration is an INDEPENDENT flag, mirroring the proto's
    ``PropertyDescriptor``. There is deliberately no ``kind`` discriminator:
    a single mutually-exclusive axis is what made declarations non-composable
    and produced the same bug four times. ``relation_kind`` is the one
    exception — relations serialize to ``RelationDescriptor``, a different
    message, so a field is either a scalar property or a relation.
    """
    key: bool = False
    search_key: bool = False
    search_key_order: int = 0
    large_field: bool = False
    embedding: bool = False
    chunk: bool = False
    chunk_max_tokens: int = 512
    chunk_overlap: int = 64
    chunk_contextual: bool = False
    metadata: bool = False
    summary: bool = False
    keywords: bool = False
    extract_hint: str | None = ""
    description: str = ""
    relation_kind: str | None = None
    related_type: str | None = None


# ── Public factory helpers ─────────────────────────────────────────────────────

def iverson_field(
    *,
    key: bool = False,
    search_key: bool = False,
    search_key_order: int = 0,
    large_field: bool = False,
    embedding: bool = False,
    chunk: bool = False,
    chunk_max_tokens: int = 512,
    chunk_overlap: int = 64,
    chunk_contextual: bool = False,
    metadata: bool = False,
    summary: bool = False,
    keywords: bool = False,
    extract_hint: str | None = "",
    description: str = "",
) -> FieldMeta:
    """Declare a field with any combination of scalar declarations.

    This is the single constructor for scalar declarations; the named
    factories below are one-line presets over it. Use this directly whenever
    a field carries MORE THAN ONE declaration.

    Args:
        key: primary key field.
        search_key / search_key_order: StarRocks MV search/sort key and its
            0-based position in the composite key.
        large_field: excluded from materialized views.
        embedding: source for a whole-field vector embedding.
        chunk / chunk_max_tokens / chunk_overlap / chunk_contextual: source
            for chunk-level vector embeddings and its windowing.
        metadata: a property that describes or qualifies the entity rather
            than carrying its primary content.
        summary / keywords: Ollama enrichment targets.
        extract_hint: Ollama extraction target, guided by this hint. ``""``
            means "not an extraction target"; a blank-but-non-empty hint is
            rejected.
        description: human-readable field description.
    """
    # The server treats an empty hint as "not an extraction target" and would
    # silently drop it, so a None or blank-but-non-empty hint is rejected here.
    # An empty string is the "not declared" default and must NOT raise.
    if extract_hint is None or (extract_hint != "" and not extract_hint.strip()):
        raise ValueError(
            "extract_hint must be non-blank when provided; the server treats "
            "an empty hint as \"not an extraction target\" and would silently "
            "drop it."
        )

    return FieldMeta(
        key=key,
        search_key=search_key,
        search_key_order=search_key_order,
        large_field=large_field,
        embedding=embedding,
        chunk=chunk,
        chunk_max_tokens=chunk_max_tokens,
        chunk_overlap=chunk_overlap,
        chunk_contextual=chunk_contextual,
        metadata=metadata,
        summary=summary,
        keywords=keywords,
        extract_hint=extract_hint,
        description=description,
    )


def iverson_key(description: str = "") -> FieldMeta:
    """Mark the primary key field of an entity."""
    return iverson_field(key=True, description=description)


def iverson_search_key(order: int = 0, description: str = "") -> FieldMeta:
    """Mark a field used as a search/sort key in the StarRocks MV."""
    return iverson_field(search_key=True, search_key_order=order, description=description)


def iverson_metadata(description: str = "") -> FieldMeta:
    """Mark a field as a metadata signal."""
    return iverson_field(metadata=True, description=description)


def iverson_description(description: str) -> FieldMeta:
    """Supply a human-readable description for an otherwise plain field."""
    return iverson_field(description=description)


def iverson_large_field(description: str = "") -> FieldMeta:
    """Mark a field as large (excluded from materialized views)."""
    return iverson_field(large_field=True, description=description)


def iverson_embedding(description: str = "") -> FieldMeta:
    """Mark a string field as a source for a whole-field vector embedding."""
    return iverson_field(embedding=True, description=description)


def iverson_chunk(
    max_tokens: int = 512,
    overlap: int = 64,
    description: str = "",
    contextual: bool = False,
) -> FieldMeta:
    """Mark a string field as a source for chunk-level vector embeddings."""
    return iverson_field(
        chunk=True,
        chunk_max_tokens=max_tokens,
        chunk_overlap=overlap,
        chunk_contextual=contextual,
        description=description,
    )


def iverson_summary(description: str = "") -> FieldMeta:
    """Mark a field as the target for an Ollama-driven summary."""
    return iverson_field(summary=True, description=description)


def iverson_keywords(description: str = "") -> FieldMeta:
    """Mark a field as the target for Ollama-driven keyword extraction."""
    return iverson_field(keywords=True, description=description)


def iverson_extracted(hint: str, description: str = "") -> FieldMeta:
    """Mark a field as an Ollama extraction target, guided by ``hint``.

    The hint is mandatory here, unlike the optional ``extract_hint`` kwarg on
    ``iverson_field`` where ``""`` means "not declared". Normalizing ``""`` to
    ``None`` routes an empty hint into the shared guard's rejection path.
    """
    return iverson_field(extract_hint=hint or None, description=description)


def many_to_one(type_name: str) -> FieldMeta:
    """Declare a many-to-one relation field (FK on this entity)."""
    return FieldMeta(relation_kind="many_to_one", related_type=type_name)


def many_to_many(type_name: str) -> FieldMeta:
    """Declare a many-to-many relation field."""
    return FieldMeta(relation_kind="many_to_many", related_type=type_name)


def one_to_many(type_name: str) -> FieldMeta:
    """Declare a one-to-many relation field (FK on the related entity)."""
    return FieldMeta(relation_kind="one_to_many", related_type=type_name)


def one_to_one(type_name: str) -> FieldMeta:
    """Declare a one-to-one relation field."""
    return FieldMeta(relation_kind="one_to_one", related_type=type_name)


# ── Entity registry ─────────────────────────────────────────────────────────────

#: Maps ``cls.__name__`` (the same string ``_iverson_meta["type_name"]`` carries,
#: and the same string a relation's ``related_type`` holds) to the decorated
#: class itself. Populated at import time by ``iverson_entity``, before any read
#: can occur, so the read-path hydration pass can resolve a relation's
#: ``related_type`` back to a concrete class.
ENTITY_REGISTRY: dict[str, type] = {}


# ── @iverson_entity decorator ──────────────────────────────────────────────────


def iverson_entity(cls: type | None = None, *, description: str = "", embedding_model: str | None = None):
    """Class decorator that collects ``FieldMeta`` annotations into metadata.

    Usable bare (``@iverson_entity``) or called
    (``@iverson_entity(description="...")`` to describe the type itself).

    Args:
        description: human-readable description of the type itself.
        embedding_model: the Ollama model this type's embedding/chunk properties are
            generated with. One model per TYPE, never per property — every
            ``iverson_embedding()``/``iverson_chunk()`` field on this class is stamped with
            the same value. Three cases: ``None`` (the default, i.e. the argument was not
            supplied) means "not declared HERE" — if a decorated base class in this class's MRO
            declares a non-empty model, the nearest such declaration is inherited. An explicit
            ``""`` means "declared HERE as opted OUT" — no MRO walk happens, and this type's
            properties are stamped with ``""`` even if a base declares a model, matching how the
            other Iverson clients treat an explicit empty value. A non-empty value is used as-is
            and overrides any inherited one. If resolution (supplied or inherited) ends in ``""``
            or ``None`` with no declaring base, the server resolves the deployment's configured
            default, exactly as it does for a client that predates this parameter — so an
            un-updated caller keeps working with no server-side special-casing. Only
            ``embedding_model`` inherits this way from a decorated base — its ``FieldMeta``
            sentinels (including ``iverson_key()``) are replaced with ``None`` on the base (see
            "After decoration" below) and do NOT carry into subclasses, so the declaring base
            must be field-less, or the child must redeclare every field it needs.

    After decoration:
    - ``cls._iverson_meta`` is a dict with keys:
        - ``type_name`` (str): simple class name
        - ``key_field`` (str | None): name of the ``iverson_key()`` field
        - ``search_keys`` (list[tuple[str, int]]): [(field_name, order), ...]
        - ``large_fields`` (list[str]): field names marked iverson_large_field
        - ``relations`` (list[dict]): each dict has 'field', 'kind', 'related_type'
        - ``fields`` (list[str]): all annotated field names (key + plain + large)
        - ``metadata_fields`` (list[str]): field names marked as metadata signals
        - ``descriptions`` (dict[str, str]): field name → description text
        - ``description`` (str): description of the type itself
        - ``embedding_model`` (str): the declared embedding model, or ``""`` if undeclared
    - Every ``FieldMeta`` class attribute is replaced with ``None`` so instances
      can set it normally.
    """
    if cls is None:
        def _decorate(inner_cls: type) -> type:
            return iverson_entity(
                inner_cls, description=description, embedding_model=embedding_model,
            )
        return _decorate

    annotations: dict[str, Any] = {}
    # Walk MRO to gather inherited annotations (excluding object)
    for base in reversed(cls.__mro__):
        if base is object:
            continue
        annotations.update(getattr(base, "__annotations__", {}))

    if embedding_model is None:
        # None (not supplied) means "inherit from the MRO"; an explicit "" is a deliberate
        # opt-out and must NOT walk the MRO — see the docstring's three cases. Unlike the
        # annotation gather above (which walks farthest-first to accumulate every inherited
        # field), this walks NEAREST-first — cls.__mro__[1:] is parent, then grandparent, ... —
        # so a middle class's own declaration wins over an ancestor's. Skip object and any
        # undecorated base (neither carries `_iverson_meta`). If no declaring base is found,
        # resolve to "" (the "undeclared" sentinel `_iverson_meta["embedding_model"]` carries).
        embedding_model = ""
        for base in cls.__mro__[1:]:
            base_meta = getattr(base, "_iverson_meta", None)
            if base_meta and base_meta.get("embedding_model"):
                embedding_model = base_meta["embedding_model"]
                break

    key_field: str | None = None
    search_keys: list[tuple[str, int]] = []
    large_fields: list[str] = []
    embedding_fields: list[str] = []
    chunk_fields: list[tuple[str, int, int, bool]] = []
    relations: list[dict] = []
    plain_fields: list[str] = []
    metadata_fields: list[str] = []
    descriptions: dict[str, str] = {}
    summary_fields: list[str] = []
    keywords_fields: list[str] = []
    extracted_fields: dict[str, str] = {}

    for field_name, _type_hint in annotations.items():
        default = getattr(cls, field_name, None)
        if isinstance(default, FieldMeta):
            meta: FieldMeta = default
            # Replace the FieldMeta sentinel with None so the attribute is usable
            setattr(cls, field_name, None)

            if meta.relation_kind:
                # Relations are the one exclusive axis: they serialize to
                # RelationDescriptor, not PropertyDescriptor. The dict keeps its
                # "kind" key because core.py:69 and core.py:188 read rel["kind"].
                relations.append({
                    "field": field_name,
                    "kind": meta.relation_kind,
                    "related_type": meta.related_type,
                })
                continue

            if meta.key:
                key_field = field_name
            if meta.search_key:
                search_keys.append((field_name, meta.search_key_order))
            if meta.large_field:
                large_fields.append(field_name)
            if meta.embedding:
                embedding_fields.append(field_name)
            if meta.chunk:
                chunk_fields.append(
                    (field_name, meta.chunk_max_tokens, meta.chunk_overlap, meta.chunk_contextual))
            if meta.metadata:
                metadata_fields.append(field_name)
            if meta.summary:
                summary_fields.append(field_name)
            if meta.keywords:
                keywords_fields.append(field_name)
            if meta.extract_hint:
                extracted_fields[field_name] = meta.extract_hint
            if meta.description:
                descriptions[field_name] = meta.description

            # Exactly one append per scalar field. Previously this happened inside
            # each elif branch AND in the terminal else; under independent ifs that
            # would double-append any field carrying two flags, and core.py:147
            # iterates this list directly — a duplicate emits a duplicate
            # PropertyDescriptor on the wire.
            plain_fields.append(field_name)
        else:
            plain_fields.append(field_name)

    search_keys.sort(key=lambda t: t[1])

    cls._iverson_meta = {
        "type_name": cls.__name__,
        "key_field": key_field,
        "search_keys": search_keys,
        "large_fields": large_fields,
        "embedding_fields": embedding_fields,
        "chunk_fields": chunk_fields,
        "relations": relations,
        "fields": plain_fields,
        "metadata_fields": metadata_fields,
        "descriptions": descriptions,
        "description": description,
        "embedding_model": embedding_model,
        "summary_fields": summary_fields,
        "keywords_fields": keywords_fields,
        "extracted_fields": extracted_fields,
    }

    ENTITY_REGISTRY[cls.__name__] = cls

    return cls
